sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/DefaultStargateBridgeClient.java

```diff
@@ -76,120 +82,172 @@
+  public CompletionStage<Response> executeBatchAsync(Batch batch) {
+    ClientCall<Batch, Response> call =
+        channel.newCall(StargateBridgeGrpc.getExecuteBatchMethod(), callOptions);
+    UnaryStreamObserver<Response> observer = new UnaryStreamObserver<>();
+    ClientCalls.asyncUnaryCall(call, batch, observer);
+    return observer.asFuture();
   }
 
   @Override
-  public Optional<CqlKeyspaceDescribe> getKeyspace(String keyspaceName)
-      throws UnauthorizedKeyspaceException {
-    if (!authorizeSchemaRead(SchemaReads.keyspace(keyspaceName, sourceApi))) {
-      throw new UnauthorizedKeyspaceException(keyspaceName);
-    }
-    return getAuthorizedKeyspace(keyspaceName);
+  public CompletionStage<Optional<CqlKeyspaceDescribe>> getKeyspaceAsync(String keyspaceName) {
+    return authorizeSchemaReadAsync(SchemaReads.keyspace(keyspaceName, sourceApi))
+        .thenApply(
+            authorized -> {
+              if (!authorized) {
+                throw new UnauthorizedKeyspaceException(keyspaceName);
+              }
+              return authorized;
+            })
+        .thenCompose(__ -> getAuthorizedKeyspace(keyspaceName));
   }
 
-  private Optional<CqlKeyspaceDescribe> getAuthorizedKeyspace(String keyspaceName) {
-    return Optional.ofNullable(
-        stripTenantPrefix(schema.getKeyspace(addTenantPrefix(keyspaceName))));
+  private CompletionStage<Optional<CqlKeyspaceDescribe>> getAuthorizedKeyspace(
+      String keyspaceName) {
+    return schema
+        .getKeyspaceAsync(addTenantPrefix(keyspaceName))
+        .thenApply(ks -> Optional.ofNullable(stripTenantPrefix(ks)));
   }
 
   @Override
-  public List<CqlKeyspaceDescribe> getAllKeyspaces() {
-    List<String> names = getKeyspaceNames(new ArrayList<>(), null);
-    List<Boolean> authorizations =
-        authorizeSchemaReads(
-            names.stream()
-                .map(n -> SchemaReads.keyspace(n, sourceApi))
-                .collect(Collectors.toList()));
-    List<CqlKeyspaceDescribe> keyspaces = new ArrayList<>(names.size());
-    int i = 0;
-    for (String name : names) {
+  public CompletionStage<List<CqlKeyspaceDescribe>> getAllKeyspacesAsync() {
+    CompletionStage<List<String>> allNamesFuture = getKeyspaceNames(new ArrayList<>(), null);
+    CompletionStage<List<String>> authorizedNamesFuture =
+        allNamesFuture.thenCompose(
+            names -> {
+              List<SchemaRead> reads =
+                  names.stream()
+                      .map(n -> SchemaReads.keyspace(n, sourceApi))
+                      .collect(Collectors.toList());
+              return authorizeSchemaReadsAsync(reads)
+                  .thenApply(authorizations -> removeUnauthorized(names, authorizations));
+            });
+    return authorizedNamesFuture.thenCompose(this::getAuthorizedkeyspaces);
+  }
+
+  private <T> List<T> removeUnauthorized(List<T> elements, List<Boolean> authorizations) {
+    assert elements.size() == authorizations.size();
```

Comment:
throw proper exception and msg instead of `assert`?

Reply:
I guess it depends how we decide to use (or not use) assertions... Here it's intended more as documenting an invariant of the method: `authorizations` is always the result of a call to `authorizeSchemaReadsAsync`, which returns a list of the same size as its argument. So different sizes cannot really happen, unless there is a programming error.
If I didn't use an assertion here, I would probably just remove the statement altogether.